catalog/controller/extension/payment/cardinity.php

```diff
@@ -143,68 +139,93 @@ private function getSession($sessionId)
+            }else{
+                $this->model_extension_payment_cardinity->log("Payment finalized by notification");
+            }
+        } else {
+            $this->model_extension_payment_cardinity->addOrder(array(
+                'order_id'   => $_POST['order_id'],
+                'payment_id' => $_POST['id'],
+            ));
+            $this->model_extension_payment_cardinity->updateOrder(array(
+                'payment_status' => 'failed_external',
+                'payment_id' => $_POST['id'],
+            ));
+
+            $this->failedOrder("Card was declined", $this->language->get("error_payment_declined"));
+
+            if($redirect){
+                $this->response->redirect($this->url->link('checkout/checkout', '', true));
+            }
+        }
+    }
 	public function externalPaymentCallback()
 	{
-		$this->load->language('extension/payment/cardinity');
-		$this->load->model('extension/payment/cardinity');
-
-		//restore session based on sessionId from description
-		$this->session->start($_POST['description']);
-		$this->getSession($_POST['description']);
-
-		$message = '';
-		ksort($_POST);
-
-		foreach ($_POST as $key => $value) {
-			if ($key == 'signature') continue;
-			$message .= $key . $value;
-		}
-
-		error_reporting(null);
-		$signature = hash_hmac('sha256', $message, $this->config->get('payment_cardinity_project_secret'));
-		error_reporting(E_ALL);
-
-		if ($signature == $_POST['signature'] && $_POST['status'] == 'approved') {
-
-			$this->model_extension_payment_cardinity->addOrder(array(
-				'order_id'   => $_POST['order_id'],
-				'payment_id' => $_POST['id'],
-			));
-			$this->model_extension_payment_cardinity->updateOrder(array(
-				'payment_status' => 'approved_external',
-				'payment_id' => $_POST['id'],
-			));
-
-			$this->logTransaction(array(
-				'orderId' => $_POST['order_id'],
-				'transactionId' =>  $_POST['id'],
-				'3dsVersion' => 'unknown (external)',
-				'amount' => $_POST['amount'] . " " . $_POST['currency'],
-				'status' => 'approved'
-			));
-
-			$this->finalizeOrder($_POST);
-			$this->response->redirect($this->url->link('checkout/success', '', true));
-		} else {
-
-			$this->model_extension_payment_cardinity->addOrder(array(
-				'order_id'   => $_POST['order_id'],
-				'payment_id' => $_POST['id'],
-			));
-			$this->model_extension_payment_cardinity->updateOrder(array(
-				'payment_status' => 'failed_external',
-				'payment_id' => $_POST['id'],
-			));
-
-			$this->failedOrder("Card was declined", $this->language->get("error_payment_declined"));
-			$this->response->redirect($this->url->link('checkout/checkout', '', true));
-		}
+        $this->finishProcessExternalPayment(true);
 	}
 
+    public function externalPaymentNotification()
+    {
+        $this->finishProcessExternalPayment(false);
+        echo json_encode([
+            "message" => "Oc3 received notification",
```

Comment:
Is this message printed in logs, or is it visible to cardholders?

Reply:
its returned to cardinity hosted payment. and hosted payment logs this. Helps debug notification on aws.